### ISSUE TITLE
Why does the ONNX model exported from PyTorch have much lower accuracy than the .pt model?

### ISSUE THREAD

**Issue by thomas7-24** [ultralytics/ultralytics#24893]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

python           3.10
torch                    2.10.0+cu130     
torchaudio            2.10.0+cu130     
torchvision            0.25.0+cu130    
onnx                      1.22.0          
onnxruntime          1.23.2

**Comment by UltralyticsAssistant** [comment]
👋 Hello @thomas7-24, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your issue triaged quickly, and an Ultralytics engineer will also assist you soon.
We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.
Since this issue involves a discrepancy between PyTorch and ONNX model accuracy, please share as much detail as possible about your export and validation workflow. If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) to help us debug it.
If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).
Helpful details may include the exact export command, inference command or script used for both `.pt` and ONNX, any preprocessing and postprocessing steps, confidence/IoU settings, image size, whether validation was run through Ultralytics or external ONNX runtime code, and a small reproducible sample if possible 🙂.
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Hello.

Please provide more information about your issue including the model you're using, the export command used and also output of `yolo checks`.

**Comment by thomas7-24** [comment]
Simple export command

def Export(modelPath):
    model = YOLO(modelPath) 
    model.export(format="onnx",
                 #dynamic=False, 
                # simplify=False,
                # half=False, 
                 #opset=18
                )

**Comment by glenn-jocher** [comment]
Thanks — that export call is the expected one, so if the `.onnx` result is much worse than the `.pt`, the cause is often the validation/inference pipeline rather than the export itself; please run the same validation on both models as shown in the [ONNX export docs](https://docs.ultralytics.com/integrations/onnx) and share the full text output from both plus `yolo checks` (or `ultralytics.utils.checks.collect_system_info()` if you may have multiple installs), and if you’re using custom ONNX Runtime inference please also post that code since preprocessing/NMS mismatches are a common cause:
```bash
yolo val model=path/to/best.pt data=path/to/data.yaml
yolo val model=path/to/best.onnx data=path/to/data.yaml
yolo checks
```

**Comment by Y-T-G** [comment]
Duplicate of https://github.com/ultralytics/ultralytics/issues/24861

**Comment by Y-T-G** [comment]
> Simple export command
> 
> def Export(modelPath):
>     model = YOLO(modelPath) 
>     model.export(format="onnx",
>                  #dynamic=False, 
>                 # simplify=False,
>                 # half=False, 
>                  #opset=18
>                 )
> 
> 

To match PyTorch output, you should export with `dynamic=True`.

**Comment by thomas7-24** [comment]
OK, thanks, I've already found the problem. The preprocessing of the inference engine is different.

**Comment by glenn-jocher** [comment]
Thanks for confirming — that’s a common root cause, and keeping resize, normalization, color order, and postprocessing identical between the `.pt` and ONNX pipelines, then doing a quick parity check like the [export verification example](https://docs.ultralytics.com/guides/export-non-yolo-models/#verify-your-exported-model), is the best way to catch it early.
